Make pipeline tasks easier to debug.

HassanJaveed84 opened the pull request:
@mulby @brianhw I added a script that pulls the emr logs and displays Traceback info from the downloaded logs.

brianhw commented:
Thanks for doing this, Hassan!  

I would be interested in knowing how performant this approach is.  What Gabe described is very similar to what I used to do, though I used my s3_util script to only pull down files ending with "stderr" to avoid pulling "stdlog" (or something) files that were big but not useful.  

With the switch to Hadoop 2 and the non-reuse of containers, I found that some of the nightly jobs generated so many stderr files (one per container) that it took a lot of time (and disk space) to do the download.  Your approach might be faster by using aws s3 sync instead, but I'd be curious. 

Another approach that would minimize local disk space use would be to get a listing of the remote files and then stream them one by one, and implement the grep either in code or on a temp file. 

A small thing:  it might be nice to occasionally override the grep parameter for after context, in those cases where 30 is not adequate.   (Seems like a fine default.)

mulby commented on edx/analytics/tasks/launchers/debug_emr_logs.py:
```diff
@@ -0,0 +1,83 @@
+
+    args = arg_parser.parse_args()
+
+    s3_emr_logs_url = "s3://{bucket_name}/{job_flow}/containers/{application_id}".format(
```
job_flow here has to be the cluster ID (for example: j-123457AC83754). I often want to just give this script the human-readable cluster name, like "Enrollment".

When the cluster is deployed we store this human readable name on it... ideally this script could use boto to find the most recently deployed cluster with that name and grab it's cluster ID to build this S3 URL.

See https://github.com/edx/edx-analytics-configuration/blob/master/batch/library/emr#L40

tobz replied:
I think it makes sense to allow both.  Give the user an option to specify the cluster name, which finds the latest job flow ID which matches the name, or let them specify the job flow ID specifically.

If the cluster name is present, it just looks it up, and the rest of the code is the same, as if they passed in a job flow ID.

The worst thing would be that we actually start to use this task because it's helpful, and then we want to go look at the error logs for the 2nd to last run of the workflow, and it can only show us the latest one. :P

brianhw replied:
Amen to the last point -- it may often be that we do the analysis of why a job failed after it has been started a second time.  

I'm imagining down the road adding a build step that runs to do this dumping if a job fails as part of the definition on Jenkins.  For that we might still have the cluster running, so asking for the latest that had completed might not quite be the right one.  But we would hopefully get access to the current job flow id.

mulby commented on edx/analytics/tasks/launchers/debug_emr_logs.py:
```diff
@@ -0,0 +1,83 @@
+def download_emr_logs(s3_emr_logs_url, output_path):
+    print("Downloading logs.")
+    proc = Popen(
+        ['aws','s3','sync',s3_emr_logs_url, output_path],
```
Ideally, I'd prefer not to actually persist the logs to my local disk... but I'd be OK with this if needed!

We also might be able to use the "--include" or "--exclude" parameters to limit the set of files returned to just the stderr ones.

brianhw replied:
Or again, use an s3util-like mechanism to provide an outer loop for processing a single file at a time.

mulby commented on edx/analytics/tasks/launchers/debug_emr_logs.py:
```diff
@@ -0,0 +1,83 @@
+                        for line in input_file:
+                            output_file.write(line)
+
+def extract_errors(root_path):
```
We could probably borrow the logic that luigi uses here: https://github.com/edx/luigi/blob/edx/release/luigi/hadoop.py#L350

I think the stack trace is hex encoded by mrrunner.py here: https://github.com/edx/luigi/blob/edx/release/luigi/mrrunner.py#L63

So we could just take advantage of that fact to find them all without having a fixed window of N lines of output.

brianhw replied:
It would also be good to know what file a given traceback was found from.

mulby commented on edx/analytics/tasks/launchers/debug_emr_logs.py:
```diff
@@ -0,0 +1,83 @@
+
+def main():
+    arg_parser = argparse.ArgumentParser(description='Get Traceback information from emr-logs.')
+    arg_parser.add_argument(
```
It would be great if this defaulted to a value from some kind of environment variable so I could throw it into my .bashrc and never think about it again.

tobz replied:
Yep.

HassanJaveed84 commented:
- Added option to specify bucket name via env. variable. 
- Added option to cleanup downloaded logs.
- Pull cluster id from latest cluster if cluster id is not passed.
- Used luigi's error lookup.

@mulby @brianhw @tobz

tobz commented:
We can ignore code coverage check results for now, because the thresholds haven't been set at all.
